Add Vertx.io based API with swagger UI

kismob opened the pull request:
This PR resolves https://github.com/linkedin/cruise-control/issues/1662
This task attempts to tackle the task of introducing an alternative and modern Vertex.io based API parallel to the existing servlet one and based on the existing codebase. The new API will be fully compatible with the currently existing one to ease migration.
There is 2 commits in this PR. In the first one I introduce the interfaces, and implement refactor the servlet API to implement those. This requires small changes in the Parameters and the Request classes. In the second one I implement the Vertx classes with the same interfaces.
[Here is a doc, that makes it easier to understand](https://docs.google.com/document/d/1ZedEm0IgkCUPc1Y33g4OTGCFe-ZUhqqb/edit?usp=sharing&ouid=106747861335666409026&rtpof=true&sd=true)

I know that this is a huge change, so if you would like to organise an online interactive review, we are open to that.

wyuka commented:
I am not clear on the motivation behind integrating Vert.x in Cruise Control. What is the problem it is trying to solve? The document attached has a single line about it - "Cruise Control uses the classic servlet technology which is cumbersome to use". What does that actually mean? Can you provide more details on the problems that it currently has?

Cruise Control currently exposes a REST-ful API on top of HTTP to perform operations on it. As far as I understand, that part is not going to change. In that case, who is going to benefit from this?

viktorsomogyi commented:
Hi @wyuka, the benefit that we're getting from using a framework are the following:
- we'll keep the API as it is but it'll be easier to add new functionality later on, add security features and so on
- defining the API via Swagger/OpenAPI will give us the benefit of self-documentation as well besides defining the API
- we'll get a Swagger UI that's easier to use than the current implementation (our customers currently use the rest API mostly)
The reason why Vertx.io is the choice comes from an earlier agreement with @efeg. It is more lightweight than Dropwizard or Spring (in terms of boilerplate dependencies that we don't need) and it can be integrated with RestEasy as well.

Currently we tried to keep both modes of the API (servlet vs vertx) that are compatible with each other and we tried to do it with minimal changes (well it's still 159 files but most in of them there are either small changes, refactors or just moving some files). We tried to separate the whole stuff into 2 commits to make it somewhat easier to review.

If you guys think the code change is still too big and would do an interactive review or you'd just have a discussion about it we can sometime meet on Zoom or your preferred service.

CCisGG commented on cruise-control-core/src/main/java/com/linkedin/cruisecontrol/servlet/handler/Request.java:
```diff
@@ -19,11 +16,9 @@
    */
-  void handle(HttpServletRequest request, HttpServletResponse response)
-      throws IOException, ExecutionException, InterruptedException;
+  void handle(CruiseControlRequestContext handler)
```
Is this interface change backwards compatibility? Or you simply use a CruiseControlRequestContext wrapper to replace the old request/response params everywhere?

kismob replied:
We use CruiseControlRequestContext and replaced the request/response params everywhere.

CCisGG commented on cruise-control-core/src/main/java/com/linkedin/cruisecontrol/servlet/parameters/CruiseControlParameters.java:
```diff
@@ -19,13 +19,12 @@ public interface CruiseControlParameters extends CruiseControlConfigurable {
    */
-  boolean parseParameters(HttpServletResponse response);
-
+  boolean parseParameters(CruiseControlRequestContext handler);
```
The param naming here is a big confusing. From naming conversion, either the class should call something like "CruiseControlRequestHandler", or the param should call "requestContext"

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlRequestHandler.java:
```diff
@@ -0,0 +1,164 @@
+            throws Exception {
+        CruiseControlParameters parameters;
+        RequestParameterWrapper requestParameter = requestParameterFor(endPoint);
+        if (endPoint == REVIEW) {
```
Why handlePost check the endPoint with REVIEW, but handleGet check the endpoint with REVIEW_BOARD?

kismob replied:
We used it like this because, in was used in the same way in the original KafkaCruiseControlServlet.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlRequestHandler.java:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    private void handleGet(CruiseControlRequestContext handler,
```
Param name should better be "requestContext"

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServlet.java:
```diff
@@ -1,246 +0,0 @@
-/**
- * The servlet for Kafka Cruise Control.
- */
-public class KafkaCruiseControlServlet extends HttpServlet {
```
Where did you move this class to? E.g. since the change is backwards compatible, the servlet should still be kept somewhere.

CCisGG replied:
Alright, looks like it's moved to ServletRequestHandler

kismob replied:
Yes, that's right.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServletUtils.java:
```diff
@@ -164,18 +166,17 @@ public static RequestParameterWrapper requestParameterFor(EndPoint endpoint) {
    * @return The ip address of the client sending the request.
    */
-  public static String getClientIpAddress(HttpServletRequest request) {
+  public static String getClientIpAddress(CruiseControlRequestContext handler) {
```
param name should be requestContext

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServletUtils.java:
```diff
@@ -186,22 +187,20 @@ static String urlEncode(String s) throws UnsupportedEncodingException {
    * response and (2) returns null.
    */
-  static CruiseControlEndPoint getValidEndpoint(HttpServletRequest request, HttpServletResponse response, KafkaCruiseControlConfig config)
+  public static CruiseControlEndPoint getValidEndpoint(CruiseControlRequestContext handler)
```
param name should be requestContext

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServletUtils.java:
```diff
@@ -210,56 +209,43 @@ static CruiseControlEndPoint getValidEndpoint(HttpServletRequest request, HttpSe
-                                           HttpServletResponse response,
-                                           KafkaCruiseControlConfig config)
+  public static String handleUserRequestException(UserRequestException ure,
+                                                  CruiseControlRequestContext handler)
```
param name should be requestContext

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java:
```diff
@@ -210,56 +211,52 @@ protected boolean removeEldestEntry(Map.Entry<UUID, UserTaskInfo> eldest) {
    */
-  public List<OperationFuture> getOrCreateUserTask(HttpServletRequest httpServletRequest,
-                                                   HttpServletResponse httpServletResponse,
+  public List<OperationFuture> getOrCreateUserTask(CruiseControlRequestContext handler,
```
param name should be requestContext

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/AbstractAsyncRequest.java:
```diff
@@ -42,23 +41,23 @@ public AbstractAsyncRequest() {
   @Override
-  public CruiseControlResponse getResponse(HttpServletRequest request, HttpServletResponse response)
-      throws ExecutionException, InterruptedException {
+  public CruiseControlResponse getResponse(CruiseControlRequestContext handler)
```
param name should be requestContext

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/AddBrokerRequest.java:
```diff
@@ -21,7 +21,7 @@ public AddBrokerRequest() {
 
   @Override
-  protected OperationFuture handle(String uuid) {
+  public OperationFuture handle(String uuid) {
```
why those keywords change are needed?

kismob replied:
This is a leftover mistake from previous iterations, I fixed it.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/sync/AbstractSyncRequest.java:
```diff
@@ -36,12 +35,12 @@ public AbstractSyncRequest() {
   @Override
-  public CruiseControlResponse getResponse(HttpServletRequest request, HttpServletResponse response)
-      throws ExecutionException, InterruptedException {
+  public CruiseControlResponse getResponse(CruiseControlRequestContext handler)
```
param name should be requestContext

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/PartitionLoadParameters.java:
```diff
@@ -15,17 +15,7 @@
-import static com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils.PARTITION_PARAM;
-import static com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils.MIN_VALID_PARTITION_RATIO_PARAM;
-import static com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils.BROKER_ID_PARAM;
+import static com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils.*;
```
avoid using star here.

CCisGG reviewed (commented):
What are the newly added js files used for?

CCisGG commented:
Hi @kismob and @viktorsomogyi , thanks for the contribution and your clarification. I think it's reasonable to take this change to open source.

I have started to review the PR , and it would take several passes. Later if needed, I'll request for an interactive review.

kismob commented:
Hey @CCisGG, thank you for the review, I updated the PR according to that. (And also renamed every CruiseControlRequestContext object to requestHandler).
The js files are needed for the Swagger UI.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java:
```diff
@@ -902,6 +904,23 @@ public static <K, KT extends Deserializer<K>, V, VT extends Deserializer<V>> Con
+   * @throws ServletException
+   */
+  public static KafkaCruiseControlApp getCruiseControlApp(KafkaCruiseControlConfig config, Integer port, String hostname) throws ServletException {
+    if (config.getBoolean(WebServerConfig.VERTX_ENABLED_CONFIG)) {
```
Looks like it's default to false and is always non-null.

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java:
```diff
@@ -585,24 +589,24 @@ public boolean equals(Object o) {
     @Override
     public String toString() {
-      return String.format("SessionKey{_httpSession=%s,_requestUrl=%s,_queryParams=%s}", _httpSession, _requestUrl,
+      return String.format("SessionKey{_handler=%s,_requestUrl=%s,_queryParams=%s}", _session, _requestUrl,
```
should be "_session=%s"?

CCisGG commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java:
```diff
@@ -824,4 +828,5 @@ public static List<TaskState> cachedValues() {
       return Collections.unmodifiableList(CACHED_VALUES);
     }
   }
+
```
nit: remove the extra line

CCisGG reviewed (commented):
Took a second pass.

I think I'm sort of ready for take an interactive review. Please let me know how we can connect.

Besides that, may I ask some help for the verification of several endpoints still working on servlet mode?

1. state
2. kafka_cluster_state
3. rebalance dryrun=true/false
4. remove/add broker

Could you please help us to at least verify the above endpoints with the servlet mode?
(E.g. List the command you call those endpoint and some result summary of each command.) That would be very helpful. It would be also great if you can add the new query format with Vertx so that people can understand the difference from user perspective.

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlApp.java:
```diff
@@ -8,205 +8,53 @@
   }
 
-  void start() throws Exception {
+  public void start() throws Exception {
```
Yes, this is for testing purpose

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlApp.java:
```diff
@@ -8,205 +8,53 @@
-
-    MetricRegistry metricRegistry = new MetricRegistry();
-    _jmxReporter = JmxReporter.forRegistry(metricRegistry).inDomain(METRIC_DOMAIN)
-            .createsObjectNamesWith(LegacyObjectNameFactory.getInstance()).build();
```
I put that back.

kismob commented on build.gradle:
```diff
@@ -397,6 +409,20 @@ project(':cruise-control') {
+    outputDir = "$buildDir/resources/main/webroot/"
+  }
+
+  task copyOpenApiToResources(dependsOn: [generateOpenApiJson, processResources], type: Copy) {
```
This is not necessary in both places, updated it.

viktorsomogyi commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlVertxApp.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Override
+    public void start() {
```
We might want to print something like in the servlet app:
```
>> ********************************************* <<
>> Application directory            : /Users/viktorsomogyi-vass/dev/github/linkedin/cruise-control
>> REST API available on            : /kafkacruisecontrol/*
>> Web UI available on              : /*
>> Web UI Directory                 : ./cruise-control-ui/dist/
>> Cookie prefix path               : /
>> Kafka Cruise Control started on  : http://0.0.0.0:9090/
>> CORS Enabled ?                   : false
>> ********************************************* <<
```

CCisGG reviewed (changes requested):
Hi, thanks for addressing my comments! Finished 3rd pass and now I understand how everything get hooked together. Left some extra comments. 

I think we should add at least one integration test for servlet and one integration test for vertx.

kismob commented:
Hi @CCisGG,
Thx for the review. I will address your change requests, create Integration tests and record a small demo by the end of this week.

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/AbstractRequest.java:
```diff
@@ -5,62 +5,59 @@
   public void configure(Map<String, ?> configs) {
-    _servlet = (KafkaCruiseControlServlet) validateNotNull(configs.get(KAFKA_CRUISE_CONTROL_SERVLET_OBJECT_CONFIG),
-            "Kafka Cruise Control servlet configuration is missing from the request.");
+    if (configs.get(KAFKA_CRUISE_CONTROL_REQUEST_HANDLER_OBJECT_CONFIG).getClass().equals(KafkaCruiseControlRequestHandler.class)) {
```
You are right, this is not necessary.

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ProgressResult.java:
```diff
@@ -53,14 +49,6 @@ protected String getPlaintext() {
     return sb.toString();
   }
 
-  @Override
```
It is replaced to AbstractCruiseControlResponse (CruiseControlResponse interface), because of the unification.

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ProgressResult.java:
```diff
@@ -6,20 +6,16 @@
 import com.linkedin.cruisecontrol.servlet.parameters.CruiseControlParameters;
-import javax.servlet.http.HttpServletResponse;
-import java.io.IOException;
+import com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig;
```
in the constructor

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java:
```diff
@@ -48,28 +56,44 @@ static void setResponseCode(HttpServletResponse response, int code, boolean json
   }
 
-  static void writeResponseToOutputStream(HttpServletResponse response,
+  static void writeResponseToOutputStream(RoutingContext context,
```
Well we do not use this function. It is left here by mistake.

kismob commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseUtils.java:
```diff
@@ -33,7 +34,14 @@ public final class ResponseUtils {
+   * @param json Write the response in JSON or not
+   * @param config CruiseControlConfig
+   */
+  public static void setResponseCode(HttpServletResponse response, int code, boolean json, AbstractConfig config) {
```
This change is not necessary, I changed it back.

kismob commented:
Hi @CCisGG, thank you for your patience. I updated the PR as discussed. I changed the current Integration tests to run also with `vertx.enabled=true`.
I also recorded 2 small demos one with the Swagger UI (`vertx.enabled=true`) and also one with terminal (`vertx.enabled=false`). I used local Kafka during the demos. I can record more coplex use-cases if you wish, but I recommend to "play" with it a little.
[Cruise Control with Swagger demo](https://youtu.be/pjI21LyobLI)
[Cruise Control with servlet demo](https://youtu.be/0VULRxI5Z7E)
I used this [`cruisecontrol.properties`](https://drive.google.com/file/d/1XXkyLSFnxzWjhgvgqrRlj26kpNWMBLoz/view?usp=sharing) file in both demos (in the first one with `vertx.enabled=true`).

If there is anything that would make the review easier please let us know.

CCisGG commented on cruise-control/src/integrationTest/java/com/linkedin/kafka/cruisecontrol/BrokerFailureIntegrationTest.java:
```diff
@@ -47,14 +48,21 @@ public class BrokerFailureIntegrationTest extends CruiseControlIntegrationTestHa
+  private final Boolean _vertxEnabled;
 
-  public BrokerFailureIntegrationTest(Boolean kafkaBrokerFailureDetectorEnable) {
+  public BrokerFailureIntegrationTest(Boolean kafkaBrokerFailureDetectorEnable, Boolean vertxEnabled) {
```
It looks like we just passed this parameter to the test, but we didn't actually integration-test vertx requests. Otherwuse, could you point me to where do we actually test the REST requests?

kismob replied:
Well, in VertxAvailabilityIntegrationTest we test if the endpoints are responding or not. In the modified integration test if we turn on `vertx.enabled`, then the app runs with the verx implementations of the interfaces and uses them to complete the request. So we tested if we can access the endpoints and if the logic behind the requests works fine with vertx.
But if I miss something or you would prefer some other way to test it, let me know.

CCisGG commented on config/cruisecontrol.properties:
```diff
@@ -93,7 +93,7 @@ capacity.config.file=config/capacityJBOD.json
 
 # The list of goals to optimize the Kafka cluster for with pre-computed proposals -- consider using RackAwareDistributionGoal instead of RackAwareGoal in clusters with partitions whose replication factor > number of racks. The value must be a subset of the "goals" and a superset of the "hard.goals" and "self.healing.goals".
-default.goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkInboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkOutboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyz ...
+default.goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkInboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkOutboundCapacityGoal,com.linkedin.kafka.cruiseco ...
```
I don't think this change is relevant. Please revert.

kismob replied:
Changes in the cruisecontrol.properties and capacityJBOD.json files were committed by mistake.

CCisGG commented on config/cruisecontrol.properties:
```diff
@@ -102,7 +102,7 @@ goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.BrokerSetAwareGoal,com.lin
 
 # The list of supported hard goals -- consider using RackAwareDistributionGoal instead of RackAwareGoal in clusters with partitions whose replication factor > number of racks
-hard.goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkInboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkOutboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer. ...
+hard.goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkInboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkOutboundCapacityGoal,com.linkedin.kafka.cruisecontr ...
```
Please revert.

CCisGG commented on config/cruisecontrol.properties:
```diff
@@ -174,7 +174,7 @@ num.proposal.precompute.threads=1
 
 # The zookeeper connect of the Kafka cluster
-zookeeper.connect=localhost:2181/
+zookeeper.connect=localhost:2181
```
Please revert.

CCisGG commented on config/cruisecontrol.properties:
```diff
@@ -211,7 +211,7 @@ metric.anomaly.finder.class=com.linkedin.kafka.cruisecontrol.detector.KafkaMetri
 
 # The goal violation to detect -- consider using RackAwareDistributionGoal instead of RackAwareGoal in clusters with partitions whose replication factor > number of racks
-anomaly.detection.goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkInboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkOutboundCapacityGoal,com.linkedin.kafka.cruisecont ...
+anomaly.detection.goals=com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareDistributionGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkInboundCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.NetworkOutboundCapacityGoal,com.linkedin.kafk ...
```
Please revert.

CCisGG commented on config/cruisecontrol.properties:
```diff
@@ -8,7 +8,7 @@
 
 # The Kafka cluster to control.
-bootstrap.servers=localhost:9092
+bootstrap.servers=localhost:9092,localhost:9093,localhost:9094
```
Please revert.

CCisGG commented on cruise-control/src/integrationTest/java/com/linkedin/kafka/cruisecontrol/DiskFailureIntegrationTest.java:
```diff
@@ -55,6 +61,21 @@ public class DiskFailureIntegrationTest extends CruiseControlIntegrationTestHarn
+   */
+  @Parameterized.Parameters
+  public static Collection<Boolean> data() {
+    Boolean[] data = {false, false};
```
Shouldn't this be true, false?

kismob replied:
Yes, this is a stupid mistake. I will correct it. Thank you for pointing it out.

CCisGG commented on cruise-control/src/integrationTest/java/com/linkedin/kafka/cruisecontrol/ReplicaCapacityViolationIntegrationTest.java:
```diff
@@ -42,6 +47,21 @@ public class ReplicaCapacityViolationIntegrationTest extends CruiseControlIntegr
+   */
+  @Parameterized.Parameters
+  public static Collection<Boolean> data() {
+    Boolean[] data = {false, false};
```
Shouldn't this be true, false?

CCisGG commented on cruise-control/src/main/resources/yaml/endpoints/addBroker.yaml:
```diff
@@ -36,8 +36,6 @@ AddBrokerEndpoint:
           items:
             type: string
-            # default is every goal
-          example: ["RackAwareGoal", "ReplicaCapacityGoal", "ReplicaDistributionGoal"]
```
Why removing the examples?

kismob replied:
When using the generated openapi.json in swagger it fills the request parameters with values that were given in the `example` part of the yaml. So you had to delete them from the UI before you could run a request. (bc they often failed the request)

CCisGG replied:
I think it makes more sense to me to leave them there. Users can easily change the value as needed. I think it's helpful to have those examples in each request.

CCisGG commented:
btw the recordings are awesome and very helpful! Thanks!

kismob commented:
@CCisGG Thank you for the review. I updated the PR.

CCisGG commented:
Hi @kismob , I left a minor comment on not removing the examples. Otherwise looks good to me.

CCisGG commented on config/cruisecontrol.properties:
```diff
@@ -352,3 +352,6 @@ two.step.purgatory.retention.time.ms=1209600000
 two.step.purgatory.max.requests=25
+
+#Enable Vertx API with Swagger
+vertx.enabled=false
```
Sorry I missed this one. Please add a new line at EOF.

CCisGG commented:
Hi @kismob , one last minor comment. Once you fix it, I'll approve and merge this PR. Thanks a lot for the effort!

CCisGG reviewed (approved):
Thanks for the change! @kismob

kismob commented:
Hi @CCisGG, thank you for all the reviews, I know it was a big one! :)

CCisGG commented:
Hi @kismob , I see those integration failures: https://app.circleci.com/pipelines/github/linkedin/cruise-control/2029/workflows/802b6939-665f-4104-93bc-f73c78df22f2/jobs/5779

Could you take a look to see if you have seen that before?

CCisGG commented:
@kismob Another issue I observed is: gradle build generates redundant files. I'm wondering if we should add them to gitignore.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        cruise-control/src/main/resources/webroot/.openapi-generator-ignore
        cruise-control/src/main/resources/webroot/.openapi-generator/
        cruise-control/src/main/resources/webroot/README.md

CCisGG commented:
Made a PR to fixed it: https://github.com/linkedin/cruise-control/pull/1949
